File: python/convex_polygon.py

class Vector2d(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

    # Counter clockwise normal
    def normal(self):
        return Vector2d(-self.y, self.x)

    def dot(self, v):
        assert(type(v) == Vector2d)
        return self.x * v.x + self.y * v.y

    def __repr__(self):
        return "<{}, {}>".format(self.x, self.y)

# As always, counterclockwise.
class HalfSpace:
    def __init__(self, start, end):
        assert(type(start) == tuple)
        assert(len(start) >= 2)
        assert(type(end) == tuple)
        assert(len(end) >= 2)
        self.point = start
        v = Vector2d(end[0] - start[0], end[1] - start[1])
        self.alt_point = end
        self.normal = v.normal()

    def contains(self, p):
        assert(type(p) == tuple)
        assert(len(p) >= 2)
        test = Vector2d(p[0] - self.point[0],
          p[1] - self.point[1])
        return self.normal.dot(test) >= 0

    def intersection(self, start, end):
        assert(type(start) == tuple)
        assert(len(start) >= 2)
        assert(type(end) == tuple)
        assert(len(end) >= 2)
        v0 = Vector2d(start[0] - self.point[0],
          start[1] - self.point[1])
        v1 = Vector2d(end[0] - self.point[0],
          end[1] - self.point[1])
        q = v0.dot(self.normal)
        r = v1.dot(self.normal)
        if q-r == 0:
            return None
        t = 1.*q / (q-r)
        v = Vector2d(end[0] - start[0], end[1] - start[1])
        return (start[0] + t*v.x, start[1] + t*v.y)

# I got the area equation from here:
# https://www.mathwords.com/a/area_convex_polygon.htm

# Important is this assumes a counterclockwise convention for positive
# area. If you reverse the ordering, the areas will be negative.
class ConvexPolygon:
    def __init__(self, vertices):
        assert(type(vertices) == list)
        assert(all(map(lambda x: type(x) == tuple, vertices)))
        assert(all(map(lambda x: len(x) >= 2, vertices)))
        self.vertices = vertices

    def edge_count(self):
        return len(self.vertices)

    def get_vertex(self, i):
        return self.vertices[i]

    def area(self):
        pos_diag_sum = 0.0
        neg_diag_sum = 0.0
        m = len(self.vertices)
        v = self.vertices
        for i in range(m):
            k = i + 1
            if k >= m:
                k = 0
            pos_diag_sum += v[i][0] * v[k][1]
            neg_diag_sum += v[i][1] * v[k][0]
        return (pos_diag_sum - neg_diag_sum) / 2.0

    def trim(self, start, end):
        assert(type(start) == tuple)
        assert(len(start) >= 2)
        assert(type(end) == tuple)
        assert(len(end) >= 2)
        h = HalfSpace(start, end)
        in_v = self.vertices
        out_v = []
        for i in range(len(in_v)):
            cur_point = in_v[i]
            if i == 0:
                prev_point = in_v[-1]
            else:
                prev_point = in_v[i-1]
            if h.contains(cur_point):
                if not h.contains(prev_point):
                    out_v.append(h.intersection(prev_point, cur_point))
                out_v.append(cur_point)
            elif h.contains(prev_point):
                out_v.append(h.intersection(prev_point, cur_point))
        self.vertices = out_v

File: cpp/src/convex_polygon.cpp
#include "convex_polygon.h"
#include <cassert>
#include <sstream>

Vector2d Vector2d::normal() const
{
  return Vector2d(-y_, x_);
}

double Vector2d::dot(const Vector2d& v) const
{
  return x_ * v.x_ + y_ * v.y_;
}

Vector2d Vector2d::scaled_by(double s) const
{
  return Vector2d(s * x_, s * y_);
}

bool Vector2d::operator==(const Vector2d& v) const
{
  return x_ == v.x_ && y_ == v.y_;
}

std::string Vector2d::to_string() const
{
  std::stringstream ss;
  ss << "v<" << x_ << ", " << y_ << '>';
  return ss.str();
}

Vector2d Point2d::minus(const Point2d& p) const
{
  return Vector2d(x_ - p.x_, y_ - p.y_);
}

Point2d Point2d::plus(const Vector2d& v) const
{
  return Point2d(x_ + v.x(), y_ + v.y());
}

bool Point2d::operator==(const Point2d& p) const
{
  return x_ == p.x_ && y_ == p.y_;
}

bool Point2d::operator<(const Point2d& p) const
{
  if (y_ < p.y_)
    return true;
  else if (y_ > p.y_)
    return false;
  else if (x_ < p.x_)
    return true;
  else
    return false;
}

std::string Point2d::to_string() const
{
  std::stringstream ss;
  ss << "p(" << x_ << ", " << y_ << ')';
  return ss.str();
}

HalfSpace::HalfSpace(const Point2d& p0, const Point2d& p1)
: point_(p0),
  normal_(p1.minus(p0).normal())
{ }

bool HalfSpace::contains(const Point2d& p) const
{
  const Vector2d test_v = p.minus(point_);
  return normal_.dot(test_v) >= 0;
}

Point2d HalfSpace::intersection(const Point2d& start, const Point2d& end) const
{
  Vector2d v0 = start.minus(point_);
  Vector2d v1 = end.minus(point_);
  const double q = v0.dot(normal_);
  const double r = v1.dot(normal_);
  // For now, I'm just going to make an assert:
  assert(q - r != 0.0);
  const double t = q / (q - r);
  const Vector2d v = end.minus(start);
  return start.plus(v.scaled_by(t));
}

std::string HalfSpace::to_string() const
{
  std::stringstream ss;
  ss << "hs{ " << point_.to_string() << ", " << normal_.to_string() << " }";
  return ss.str();
}

ConvexPolygon::ConvexPolygon(const std::vector<Point2d>& vertices)
: vertices_(vertices)
{
  assert(vertices.size() > 2);
  for (std::size_t i = 0; i < vertices.size(); ++i) {
    std::size_t next_i = i + 1;
    if (next_i == vertices.size())
      next_i = 0;
    halfspaces_.emplace_back(vertices[i], vertices[next_i]);
  }
}

std::size_t ConvexPolygon::edge_count() const
{
  return vertices_.size();
}

const Point2d& ConvexPolygon::get_vertex(std::size_t i) const
{
  return vertices_[i];
}

double ConvexPolygon::area() const
{
  double positive_diagonal_sum = 0;
  double negative_diagonal_sum = 0;
  for (std::size_t i = 0; i < vertices_.size(); ++i) {
    std::size_t k = i + 1;
    if (k == vertices_.size())
      k = 0;
    positive_diagonal_sum += vertices_[i].x() * vertices_[k].y();
    negative_diagonal_sum += vertices_[i].y() * vertices_[k].x();
  }

  return (positive_diagonal_sum - negative_diagonal_sum) / 2.0;
}

void ConvexPolygon::trim(const Point2d& start, const Point2d& end)
{
  const HalfSpace h(start, end);
  std::vector<Point2d> new_vertices;
  for (std::size_t i = 0; i < vertices_.size(); ++i) {
    const Point2d& current_point = vertices_[i];
    const Point2d& previous_point =
      i == 0 ? vertices_[vertices_.size() - 1] : vertices_[i - 1];

    if (h.contains(current_point)) {
      if (!h.contains(previous_point))
        new_vertices.push_back(h.intersection(previous_point, current_point));
      new_vertices.push_back(current_point);
    } else if (h.contains(previous_point)) {
      new_vertices.push_back(h.intersection(previous_point, current_point));
    }
  }

  vertices_ = new_vertices;
}

bool ConvexPolygon::contains(const Point2d& p) const
{
  for (const HalfSpace& h : halfspaces_)
    if (!h.contains(p))
      return false;

  return true;
}

